observability/test/run_o11y_tests.py

```diff
@@ -0,0 +1,1174 @@
+                             timeout_sec: int,
+                             commands: List[str]) -> Tuple[kubernetes_stream, str]:
+        stream_resp = kubernetes_stream(self.k8s_core_v1.connect_get_namespaced_pod_exec,
+                                        pod_name, namespace, command='/bin/bash',
+                                        stderr=True, stdin=True,
+                                        stdout=True, tty=False,
+                                        _preload_content=False)
+        target_sec = int(time.time()) + timeout_sec
+        pid = ''
+        while stream_resp.is_open():
+            stream_resp.update(timeout=500) # in ms
+            if stream_resp.peek_stdout():
+                stdout = stream_resp.read_stdout().strip()
+                if not pid:
+                    pid =  self._test_pid(stdout)
+                logger.info(stdout)
+            if stream_resp.peek_stderr():
+                stderr = stream_resp.read_stderr().strip()
+                if not pid:
+                    pid =  self._test_pid(stderr)
+                logger.error(stderr)
+            if commands:
+                c = commands.pop(0)
+                logger.info('%s:$ %s' % (pod_name, c))
+                stream_resp.write_stdin(c + '\n')
+            if int(time.time()) > target_sec:
+                break
+        return stream_resp, pid
+
+    def start_testservice_server(self,
+                                 server_config: ObservabilityConfig,
+                                 timeout_secs: int,
+                                 server_start_cmd: str) -> None:
+        commands = [
+            self.set_config_cmd(server_config),
+            server_start_cmd,
+            'echo "PID=$!"'
+        ]
+        self.server_pod_stream, self.server_pid = self.exec_commands_in_pod(
+            self.server_pod_name, self.SERVER_NAMESPACE, timeout_secs, commands)
+
+    def kill_testservice_server(self) -> None:
+        logger.info('Killing process %s' % self.server_pid)
+        self.server_pod_stream.write_stdin('kill -9 %s\n' % self.server_pid)
+        self.server_pod_stream.close()
+
+    def run_client_commands(self,
+                            client_config: ObservabilityConfig,
+                            timeout_secs: int,
+                            commands: List[str]) -> None:
+        client_pod_stream, _ = self.exec_commands_in_pod(
+            self.client_pod_name, self.CLIENT_NAMESPACE, timeout_secs, [
+            self.set_config_cmd(client_config)
+        ] + commands)
+        client_pod_stream.close()
+
+TestCaseImplT = TypeVar('TestCaseImplT', bound='TestCaseImpl')
+CloudLoggingInterfaceT = TypeVar('CloudLoggingInterfaceT', bound='CloudLoggingInterface')
+CloudMonitoringInterfaceT = TypeVar('CloudMonitoringInterfaceT', bound='CloudMonitoringInterface')
+CloudTraceInterfaceT = TypeVar('CloudTraceInterfaceT', bound='CloudTraceInterface')
+
+class CloudLoggingInterface(unittest.TestCase):
```

Comment:
The structure of this doesn't seem quite right. I would have expected `TestCaseImpl` to inherit from `unittest.TestCase`, not this helper class. This class defines assertions, not test cases.

Reply:
Agreed. Will try to correct the structure in the next round.